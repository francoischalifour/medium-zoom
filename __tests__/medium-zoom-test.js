/* eslint-env jest */
const mediumZoom = require('../src/medium-zoom')

global.requestAnimationFrame = cb => setTimeout(cb, 0)

/**
 * Returns an empty `document.body` before each test.
 */
const emptyRootBeforeEach = () => {
  const root = document.body

  beforeEach(() => {
    while (root.firstChild) {
      root.removeChild(root.firstChild)
    }
  })

  return root
}

describe('public API', () => {
  test('exports mediumZoom', () => {
    expect(mediumZoom).toBeDefined()
  })

  const zoom = mediumZoom()
  const properties = [
    'images',
    'options',
    'show',
    'hide',
    'toggle',
    'update',
    'addEventListeners',
    'detach'
  ]

  properties.forEach(property => {
    test(`exports ${property}`, () => {
      expect(zoom[property]).toBeDefined()
    })
  })
})

describe('constructor', () => {
  const root = emptyRootBeforeEach()

  test('with empty string selector throws TypeError', () => {
    expect(() => {
      mediumZoom('')
    }).toThrow(TypeError)
  })

  test('without selector attaches the zoom to scaled images', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    image1.width = 200
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom()

    expect(zoom.images).toEqual([image1])
  })

  test('with a String selector attaches the zoom', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom('img')

    expect(zoom.images).toEqual([image1, image2])
  })

  test('with a Node selector attaches the zoom', () => {
    const image = document.createElement('img')
    root.appendChild(image)

    const zoom = mediumZoom(document.querySelector('img'))

    expect(zoom.images).toEqual([image])
  })

  test('with a NodeList selector attaches the zoom', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom(document.querySelectorAll('img'))

    expect(zoom.images).toEqual([image1, image2])
  })

  test('with an HTMLCollection selector attaches the zoom', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom(document.images)

    expect(zoom.images).toEqual([image1, image2])
  })

  test('with an Array selector attaches the zoom', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom([...document.querySelectorAll('img')])

    expect(zoom.images).toEqual([image1, image2])
  })

  test('with only an option object attaches scaled images and applies options', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    const p = document.createElement('p')
    image1.width = 200
    root.appendChild(image1)
    root.appendChild(image2)
    root.appendChild(p)

    const options = {
      margin: 24,
      background: '#BADA55',
      scrollOffset: 124,
      metaClick: true
    }
    const zoom = mediumZoom(options)

    expect(zoom.images).toEqual([image1])
    expect(zoom.options).toEqual(options)
    expect(image1.className).toBe('medium-zoom-image')
    expect(image2.className).toBe('')
    expect(root).toMatchSnapshot()
  })
})

describe('options', () => {
  test('contains the correct default options', () => {
    const actual = mediumZoom().options
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true
    }

    expect(actual).toEqual(expected)
  })
})

describe('images', () => {
  const root = emptyRootBeforeEach()

  test('contains no images by default', () => {
    expect(mediumZoom().images).toEqual([])
  })

  test('contains the images matching the selector', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const images = mediumZoom('img').images

    expect(images).toEqual([image1, image2])
  })
})

describe('update', () => {
  const root = emptyRootBeforeEach()

  test('without options returns correct default options', () => {
    const zoom = mediumZoom()
    const options = zoom.update()
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with margin updates the margin option and returns all options', () => {
    const zoom = mediumZoom()
    const options = zoom.update({
      margin: 48
    })
    const expected = {
      margin: 48,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with background updates the background option, returns all options and renders correctly', () => {
    const image = document.createElement('img')
    root.appendChild(image)

    const zoom = mediumZoom('img')
    zoom.show()

    const options = zoom.update({
      background: '#000'
    })
    const expected = {
      margin: 0,
      background: '#000',
      scrollOffset: 48,
      metaClick: true
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
    expect(document.querySelector('.medium-zoom-overlay').style.backgroundColor).toBe('rgb(0, 0, 0)')
  })

  test('with scroll offset updates the scroll offset option and returns all options', () => {
    const zoom = mediumZoom()
    const options = zoom.update({
      scrollOffset: 0
    })
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 0,
      metaClick: true
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with meta click updates the meta click option and returns all options', () => {
    const zoom = mediumZoom()
    const options = zoom.update({
      metaClick: false
    })
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: false
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with container as an Element updates the container option and returns all options', () => {
    const zoom = mediumZoom()
    const options = zoom.update({
      container: document.body
    })
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true,
      container: document.body
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with container as a string updates the container option and returns all options', () => {
    const zoom = mediumZoom()
    const options = zoom.update({
      container: 'body'
    })
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true,
      container: 'body'
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with container as an object updates the container option and returns all options', () => {
    const zoom = mediumZoom({
      container: {
        top: 64
      }
    })
    const options = zoom.update({
      container: {
        left: 64
      }
    })
    const expected = {
      margin: 0,
      background: '#fff',
      scrollOffset: 48,
      metaClick: true,
      container: {
        top: 64,
        left: 64
      }
    }

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })

  test('with all options updates all options and returns all options', () => {
    const expected = {
      margin: 24,
      background: '#BADA55',
      scrollOffset: 124,
      metaClick: true
    }
    const zoom = mediumZoom()
    const options = zoom.update(expected)

    expect(options).toEqual(expected)
    expect(zoom.options).toEqual(expected)
  })
})

describe('show', () => {
  const root = emptyRootBeforeEach()

  test('renders correctly', () => {
    const image = document.createElement('img')
    root.appendChild(image)

    const zoom = mediumZoom('img')
    zoom.show()

    expect(image.className).toBe('medium-zoom-image')
    expect(document.querySelector('.medium-zoom-image--open')).toBeTruthy()
    expect(document.querySelector('.medium-zoom-overlay')).toBeTruthy()
    expect(root).toMatchSnapshot()
  })

  test('renders correctly with background option', () => {
    const image = document.createElement('img')
    root.appendChild(image)

    const zoom = mediumZoom('img', {
      background: '#BADA55'
    })
    zoom.show()

    expect(root).toMatchSnapshot()
  })

  test('renders correctly with template option as an HTML Element', () => {
    const template = document.createElement('template')
    const image = document.createElement('img')
    root.appendChild(template)
    root.appendChild(image)

    const zoom = mediumZoom('img', {
      template
    })
    zoom.show()

    expect(root).toMatchSnapshot()
  })

  test('renders correctly with template option as a string', () => {
    const template = document.createElement('template')
    template.id = 'zoom-template'
    const image = document.createElement('img')
    root.appendChild(template)
    root.appendChild(image)

    const zoom = mediumZoom('img', {
      template: '#zoom-template'
    })
    zoom.show()

    expect(root).toMatchSnapshot()
  })
})

describe('hide', () => {
  const root = emptyRootBeforeEach()

  test('renders correctly', () => {
    const image = document.createElement('img')
    root.appendChild(image)

    const zoom = mediumZoom('img')
    zoom.hide()
    const classNames = [...image.classList]

    expect(classNames).toEqual(['medium-zoom-image'])
    expect(root).toMatchSnapshot()
  })
})

describe('detach', () => {
  const root = emptyRootBeforeEach()

  test('frees all images from the zoom', () => {
    const image1 = document.createElement('img')
    const image2 = document.createElement('img')
    root.appendChild(image1)
    root.appendChild(image2)

    const zoom = mediumZoom('img')
    zoom.detach()

    expect(zoom.images).toHaveLength(0)
    expect(image1.classList).toHaveLength(0)
    expect(image2.classList).toHaveLength(0)
  })
})
