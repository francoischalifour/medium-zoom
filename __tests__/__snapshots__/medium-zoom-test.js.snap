// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor with only an option object attaches scaled images and applies options 1`] = `
<body>
  <img
    class="medium-zoom-image"
    width="200"
  />
  <img />
  <p />
</body>
`;

exports[`hide renders correctly 1`] = `
<body>
  <img
    class="medium-zoom-image"
  />
</body>
`;

exports[`show renders correctly 1`] = `
<body>
  <img
    class="medium-zoom-image"
    style="visibility: hidden;"
  />
  <div
    class="medium-zoom-overlay"
    style="background-color: rgb(255, 255, 255);"
  />
  <img
    class="medium-zoom-image medium-zoom-image--open"
    style="position: absolute; top: 0px; left: 0px; width: 0px; height: 0px;"
  />
</body>
`;

exports[`show renders correctly with background option 1`] = `
<body>
  <img
    class="medium-zoom-image"
    style="visibility: hidden;"
  />
  <div
    class="medium-zoom-overlay"
    style="background-color: rgb(186, 218, 85);"
  />
  <img
    class="medium-zoom-image medium-zoom-image--open"
    style="position: absolute; top: 0px; left: 0px; width: 0px; height: 0px;"
  />
</body>
`;

exports[`show renders correctly with template option as a string 1`] = `
<body>
  <template
    id="zoom-template"
  />
  <img
    class="medium-zoom-image"
    style="visibility: hidden;"
  />
  <div
    class="medium-zoom-overlay"
    style="background-color: rgb(255, 255, 255);"
  />
  <div />
  <img
    class="medium-zoom-image medium-zoom-image--open"
    style="position: absolute; top: 0px; left: 0px; width: 0px; height: 0px;"
  />
</body>
`;

exports[`show renders correctly with template option as an HTML Element 1`] = `
<body>
  <template />
  <img
    class="medium-zoom-image"
    style="visibility: hidden;"
  />
  <div
    class="medium-zoom-overlay"
    style="background-color: rgb(255, 255, 255);"
  />
  <div />
  <img
    class="medium-zoom-image medium-zoom-image--open"
    style="position: absolute; top: 0px; left: 0px; width: 0px; height: 0px;"
  />
</body>
`;
